pkg_resources/__init__.py

```diff
@@ -1118,7 +1119,7 @@ def markers_pass(self, req: Requirement, extras: tuple[str, ...] | None = None):
         evaluation. Otherwise, return True.
         """
         extra_evals = (
-            req.marker.evaluate({'extra': extra})
+            req.marker.evaluate({'extra': extra})  # type: ignore
             for extra in self.get(req, ()) + (extras or (None,))
         )
         return not req.marker or any(extra_evals)
```

Comment:
```suggestion
        if not req.marker:
            return True
        extra_evals = (
            req.marker.evaluate({'extra': extra})
            for extra in self.get(req, ()) + (extras or ("",))
        )
        return any(extra_evals)
```
or more compact:
```suggestion
        return not req.marker or any(
            req.marker.evaluate({'extra': extra})
            for extra in self.get(req, ()) + (extras or ("",))
        )
```


The early `req.marker` check solves [union-attr](https://mypy.readthedocs.io/en/latest/_refs.html#code-union-attr).

The `None`-->`""` change solves [dict-item](https://mypy.readthedocs.io/en/latest/_refs.html#code-dict-item) and avoids a deprecated code path (see  https://github.com/pypa/packaging/blob/793ee28b4c70886e8de5731e24e444388916b3ee/src/packaging/markers.py#L322-L323)
I thought about replacing `("",)` with an empty tuple `()` since `evaluate` defaults `extra` to an empty string, but that would change the result of `any` if `self.get(req, ()) + (extras or ())` results in an empty tuple.

Reply:
I very much prefer the more compact single expression and return value.

I wonder if there's a bug reported to mypy about the `union-attr` not being resolved ahead of the lazily-generated `extra_evals` (it theoretically could detect that the generator is not entered until after the `req.marker` is resolved not to be `None`).

I'm a little uneasy with the `""`, but since that appears to be the type-spec'd indicator for the empty extra, SGTM.

Reply:
Well heck, I tried to commit the suggestion, but GitHub failed, saying that I couldn't apply changes to deleted lines and I should refresh and try again. I refreshed and tried again, but the commit message I'd created was lost. Boo GitHub.

Reply:
I tried again and got the same error:

![image](https://github.com/pypa/setuptools/assets/308610/af0732af-5cd0-4eb4-b897-3aebe22773d1)

Reply:
Indeed, when I go to the Files Changed tab, I can see that these suggestions are made against the original code, so although GitHub allows making the suggestion, it can't apply it. Also, the guidance to Refresh and try again is just broken, as anyone following that guidance would be caught in an infinite loop.

Reply:
Applied in 225f7fe32f5a07b99ff3fb403f517e6dbf208d68.

Reply:
L GitHub indeed, broken because I manually wrote `suggestion` I guess. Glad you figured it out.